### ISSUE TITLE
Server with K3S_TOKEN_FILE  and K3S_AGENT_TOKEN_FILE un-joinable.

### ISSUE THREAD

**Issue by rektide** [k3s-io/k3s#5619]
**Environmental Info:**
K3s Version: `k3s version v1.24.1-rc1+k3s1 (a5a0e8fd)go version go1.18.1

Node(s) CPU architecture, OS, and Version: `Linux epicac 5.16.0-5-amd64 #1 SMP PREEMPT Debian 5.16.14-1 (2022-03-15) x86_64 GNU/Linux`

Cluster Configuration: 1 server, 1 agent

**Describe the bug:**
Trying to run with K3S_TOKEN_FILE and K3S_AGENT_TOKEN_FILE creates only a `token` file and a `node-token` symlink (pointing at the `token` file) in the data-dir. Agents cannot connect using this token.

**Steps To Reproduce:**

I have manually copied release binaries into /usr/bin, & use systemd service files very closely resembling official services. I've specified two environment variables when launching a k3s server:
```

K3S_TOKEN_FILE="/etc/opt/k3s-yoyodyne-example/token"
K3S_AGENT_TOKEN_FILE="/etc/opt/k3s-yoyodyne-example/agent-token"
```

(I don't believe this is material but I also specify `--data-dir /var/lib/k3s-yoyodyne-example/data`.)

After launching this server, I am unable to join agent nodes to the server. Output in the next section.

As stated in the earlier section, If I look in the var/data/server directory, there is a `token` file and a `node-token` symlink that (absolutely) points at the `token` file. There are not two separate tokens. This seems unlikely to be correct.

I managed to get things working by specifying only a single token file, dropping one of the two env vars. I'll try to reproduce & test the different combinations again of what works in follow-up comments.

**Expected behavior:**
Agents should be able to join.

**Actual behavior:**

After launching this server, I am unable to join agent nodes to the server. Agents report:

```
May 28 23:26:14 epicac k3s-agent-yoyodyne-example[233995]: time="2022-05-28T23:26:14Z" level=info msg="Waiting to retrieve agent configuration; server is not ready: failed to retrieve configuration from server: https://127.0.0.1:6444/v1-k3s/config: 401 Unauthorized"
```

And on the server:

```
May 28 23:44:05 jarvis k3s-yoyodyne-example[498465]: time="2022-05-28T23:44:05Z" level=error msg="Failed to authenticate request from [2600:4040:296b:aa00::a4c]:33344: invalid username/password combination"
```

**Additional context / logs:**

Here's a listing of some key files I've used (I will delete & regenerate these secrets immediately):

```
/etc/opt/k3s-yoyodyne-example/token
oos3shaeveJohBaiquethux0Mi2achoohough1echahshahngou1wee8iiRi1yoh0eiv9Aifeeghe7IJie9ughaeku5ohgoj
/etc/opt/k3s-yoyodyne-example/agent-token
eawahsae9xu6aeweePhae2biev3aiwo6boopeefaib2ohdoe6aiJae7zeiche4eih9iesh2diu4nahqui8nuunai6uab8xao
/srv/k3s-yoyodyne-example/var/data/server/node-token
K108e239147c4ae2ce2e205bc937bb672a300bd4b0222eaa0522df722379a8aeec7::server:oos3shaeveJohBaiquethux0Mi2achoohough1echahshahngou1wee8iiRi1yoh0eiv9Aifeeghe7IJie9ughaeku5ohgoj

/srv/k3s-yoyodyne-example/var/data/server/token
K108e239147c4ae2ce2e205bc937bb672a300bd4b0222eaa0522df722379a8aeec7::server:oos3shaeveJohBaiquethux0Mi2achoohough1echahshahngou1wee8iiRi1yoh0eiv9Aifeeghe7IJie9ughaeku5ohgoj

/srv/k3s-yoyodyne-example/var/data/server/cred/passwd
eawahsae9xu6aeweePhae2biev3aiwo6boopeefaib2ohdoe6aiJae7zeiche4eih9iesh2diu4nahqui8nuunai6uab8xao,node,node,k3s:agent
oos3shaeveJohBaiquethux0Mi2achoohough1echahshahngou1wee8iiRi1yoh0eiv9Aifeeghe7IJie9ughaeku5ohgoj,server,server,k3s:server
```

One note here, the token files don't have end-of-lines. I haven't found docs for what is expected in these token files, but I believe I tested with end-of-lines & the same thing reproduced.

**Backporting**

- [ ] Needs backporting to older releases

**Comment by brandond** [comment]
The passwd file shows that the token and agent-token contents are being properly read and reflected in the file that's used to validate authentication from connecting agents, so it should be fine on the server side.

How are you passing the token to the agents when joining them to the cluster?

**Comment by rektide** [comment]
Thank you ever so much for the prompt reply @brandond.

I've been copying the `/srv/k3s-yoyodyne-example/var/data/server/token` file from the server onto the agent & setting [`K3S_TOKEN_FILE` in accordance with the agent config reference](https://rancher.com/docs/k3s/latest/en/installation/install-options/agent-config/#cluster-options). But since the `token`  and `node-token` are the same file, I think it's failing.

There's some differences between the token files and the `cred/passwd` file contents. I was expecting the `node-token` and `token` files to be what I needed to copy to other servers/agents, but if there's something else I should be doing, my apologies & thanks for the correction. I'm not sure what to do with the `cred/passwd` contents, which, as you say, do indeed indicate both are being generated.

**Comment by brandond** [comment]
The agent-token-file flag/variable is a server-only option, used to configure a separate token for use by agents. You should be copying the file at `K3S_AGENT_TOKEN_FILE` to the agent, and using that as the value of `K3S_TOKEN_FILE` on the agent.

**Comment by rektide** [comment]
Resolved. Thanks, apologies!

**Comment by ningpingfan** [comment]
"agents trying to join with the server token, which is not allowed." What happens if the agent-token and server-token are the same as shown below: Does that allow agent to join?
sudo cat /var/lib/rancher/k3s/server/cred/passwd
31fac7268dbd090a906ed50e07cd2680,node,node,k3s:agent
31fac7268dbd090a906ed50e07cd2680,server,server,k3s:server

**Comment by brandond** [comment]
yes, that is the default state.

**Comment by ningpingfan** [comment]
document is misleading: in https://docs.k3s.io/installation/airgap
"replace mynodetoken with the node token from the server typically at /var/lib/rancher/k3s/server/node-token"
But node-token is a link to token
sudo cat /var/lib/rancher/k3s/server/token
K1061d3e75477cfeb5297c9e2072a61cd5b6fb6af692459d875fc9960577a1ab39e::server:31fac7268dbd090a906ed50e07cd2680
Actually it should be something like
"replace mynodetoken with the 1st item in agent line from the server at /var/lib/rancher/k3s/server/cred/passwd"
In my case it is 31fac7268dbd090a906ed50e07cd2680. Otherwise, it will produce the same error of  "invalid username/password combination".

**Comment by brandond** [comment]
No, the document is correct. You can use the node-token/token file to join either servers or agents. The agent token can be used only to join agents. Note that if you are using agent-token, you must specify the same token AND agent-token on all servers.

If you use just the password from the passwd file (which you will note is the same as the value at the end of the long K10-format token from the other file), you bypass CA validation when joining nodes.

**Comment by ningpingfan** [comment]
OK, you are right. But if I used contents of token, I cannot pass the CA validation: Any suggestions?
========================================================
Feb 10 09:03:51  k3s[32393]: time="2023-02-10T09:03:51-05:00" level=error msg="failed to get CA certs: https://127.0.0.1:6444/cacerts: 404 Not Found"
Feb 10 09:03:53  k3s[32393]: time="2023-02-10T09:03:53-05:00" level=error msg="failed to get CA certs: https://127.0.0.1:6444/cacerts: 404 Not Found"
Feb 10 09:03:55  polkitd[895]: Unregistered Authentication Agent for unix-process:32372:23598172 (system bus name :1.1429, object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale en_US.UTF-8) (disconnected from bus)
Feb 10 09:03:55  k3s[32393]: time="2023-02-10T09:03:55-05:00" level=error msg="failed to get CA certs: https://127.0.0.1:6444/cacerts: 404 Not Found"
Feb 10 09:03:57  k3s[32393]: time="2023-02-10T09:03:57-05:00" level=error msg="failed to get CA certs: https://127.0.0.1:6444/cacerts: 404 Not Found"
Feb 10 09:03:59  k3s[32393]: time="2023-02-10T09:03:59-05:00" level=error msg="failed to get CA certs: https://127.0.0.1:6444/cacerts: 404 Not Found"
Feb 10 09:03:59  systemd[1]: cbagentd.service holdoff time over, scheduling restart.
Feb 10 09:03:59  systemd[1]: Stopped Carbon Black Predictive Security Cloud Endpoint Agent..
Feb 10 09:03:59  systemd[1]: Started Carbon Black Predictive Security Cloud Endpoint Agent..
Feb 10 09:04:00  systemd[1]: cbagentd.service: main process exited, code=exited, status=1/FAILURE
Feb 10 09:04:00  systemd[1]: Unit cbagentd.service entered failed state.
Feb 10 09:04:00  systemd[1]: cbagentd.service failed.

**Comment by brandond** [comment]
Your agent is getting a `404 Not Found` error when trying to retrieve the certs, which isn't something that would be caused by an invalid token. Check the rest of your agent configuration to make sure you don't have a typo elsewhere.
